pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+		if err != nil {
+			log.Errorf("error deserializing pubsub metadata: %s", err)
+			if dlqErr := a.sendBulkToDLQIfConfigured(ctx, psName, msg, route, nil, nil, &bulkSubDiag); dlqErr != nil {
+				populateAllBulkResponsesWithError(msg, &bulkResponses, err)
+				reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+				return bulkResponses, err
+			}
+			reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+			return nil, nil
+		}
+		routePathBulkMessageMap := make(map[string]pubsubBulkSubscribedMessage)
+		entryIDIndexMap := make(map[string]int, len(msg.Entries))
+		hasAnyError := false
+		for i, message := range msg.Entries {
+			if entryIDErr := validateEntryID(message.EntryID, i); entryIDErr != nil {
+				bulkResponses[i].Error = entryIDErr
+				hasAnyError = true
+				continue
+			}
+			entryIDIndexMap[message.EntryID] = i
+			if rawPayload {
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, string(message.Event), psName, topic, &bulkSubDiag)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
+				}
+				dataB64 := base64.StdEncoding.EncodeToString(message.Event)
+				var contenttype string
+				if message.ContentType != "" {
+					contenttype = message.ContentType
+				} else {
+					contenttype = "application/octet-stream"
+				}
+				populateBulkSubcribedMessage(&(msg.Entries[i]), dataB64, &routePathBulkMessageMap, rPath, i, msg, false, psName, contenttype)
+			} else {
+				var cloudEvent map[string]interface{}
+				err = json.Unmarshal(message.Event, &cloudEvent)
+				if err != nil {
+					log.Errorf("error deserializing one of the messages in bulk cloud event in pubsub %s and topic %s: %s", psName, msg.Topic, err)
+					bulkResponses[i].Error = err
+					bulkResponses[i].EntryID = message.EntryID
+					hasAnyError = true
+					continue
+				}
+				if pubsub.HasExpired(cloudEvent) {
+					log.Warnf("dropping expired pub/sub event %v as of %v", cloudEvent[pubsub.IDField], cloudEvent[pubsub.ExpirationField])
+					bulkSubDiag.statusWiseDiag[string(pubsub.Drop)]++
+					if route.deadLetterTopic != "" {
+						_ = a.sendToDeadLetter(psName, &pubsub.NewMessage{
+							Data:        message.Event,
+							Topic:       topic,
+							Metadata:    message.Metadata,
+							ContentType: &message.ContentType,
+						}, route.deadLetterTopic)
+					}
+					bulkResponses[i].EntryID = message.EntryID
+					bulkResponses[i].Error = nil
+					continue
+				}
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, cloudEvent, psName, topic, &bulkSubDiag)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
+				}
+				populateBulkSubcribedMessage(&(msg.Entries[i]), cloudEvent, &routePathBulkMessageMap, rPath, i, msg, true, psName, message.ContentType)
+			}
+		}
+		var overallInvokeErr error
+		for path, psm := range routePathBulkMessageMap {
```

Comment:
Once, we have iterated all messages and processed and allocated them per path, we need to send items via http/grpc per path.